More updates to testing documentation (#45528)
dotnet/docs: docs/core/testing/unit-testing-with-dotnet-test.md

### Running MTP projects with VSTest mode

> `dotnet test` is typically designed to run VSTest projects in VSTest mode, as that was its original purpose. However, to run MTP projects in `dotnet test` VSTest mode, you can use the [Microsoft.Testing.Platform.MSBuild](https://www.nuget.org/packages/Microsoft.Testing.Platform.MSBuild). From the user's perspective, this support is enabled by setting the `TestingPlatformDotnetTestSupport` MSBuild property to true (it is false by default for backward compatibility reasons). In simple terms, setting this property to true will cause Microsoft.Testing.Platform.MSBuild to change the `VSTest` target behavior, redirecting it to call `InvokeTestingPlatform`. This is an MSBuild target included in Microsoft.Testing.Platform.MSBuild, responsible for correctly running MTP test applications as executables. This means that VSTest-specific command-line options are silently ignored in this mode, such as `--logger`. This implies that there should be a way to pass MTP-specific command-line options, such as `--report-trx`, which is equivalent to using `--logger trx` in VSTest. Given the current limitations of the `dotnet test` CLI, the only way to include MTP-specific arguments is by appending them after an additional `--`. For instance, `dotnet test -- --report-trx`.

Reviewer: ```suggestion
`dotnet test` was designed to run VSTest projects in VSTest mode. However, you can run MTP projects in `dotnet test` VSTest mode by using the [Microsoft.Testing.Platform.MSBuild](https://www.nuget.org/packages/Microsoft.Testing.Platform.MSBuild) package. From the user's perspective, this support is enabled by setting the `TestingPlatformDotnetTestSupport` MSBuild property to `true` (it's `false` by default for backward-compatibility reasons). When this property is set to `true`, Microsoft.Testing.Platform.MSBuild changes the `VSTest` target behavior, redirecting it to call `InvokeTestingPlatform`. `InvokeTestingPlatform` is an MSBuild target included in Microsoft.Testing.Platform.MSBuild that's responsible for correctly running MTP test applications as executables. VSTest-specific command-line options, such as `--logger`, are silently ignored in this mode. To include MTP-specific arguments, such as `--report-trx`, you must append them after an additional `--`. For example, `dotnet test -- --report-trx`.
```

Reviewer: I tried to make it a bit less wordy.

> [!NOTE]

> > MSTest and NUnit use [Microsoft.Testing.Extensions.VSTestBridge](https://www.nuget.org/packages/Microsoft.Testing.Extensions.VSTestBridge). When you set `EnableMSTestRunner` or `EnableNUnitRunner` (the properties used to enable MTP), your test project become supporting both VSTest and Microsoft.Testing.Platform.

Reviewer 2: ```suggestion
> MSTest and NUnit use the [Microsoft.Testing.Extensions.VSTestBridge](https://www.nuget.org/packages/Microsoft.Testing.Extensions.VSTestBridge) package. By setting the `EnableMSTestRunner` or `EnableNUnitRunner` properties (which enable Microsoft.Testing.Platform), your test project will support both VSTest and Microsoft.Testing.Platform.
```

> > In that case, if you are using VSTest mode of `dotnet test` and not setting `TestingPlatformDotnetTestSupport` to true, you are actually running completely with VSTest, as if `EnableMSTestRunner` and `EnableNUnitRunner` are not there.

Reviewer 2: ```suggestion
> In that scenario, if you are using the VSTest mode of `dotnet test` and do not set `TestingPlatformDotnetTestSupport` to true, you are essentially running entirely with VSTest, as if `EnableMSTestRunner` and `EnableNUnitRunner` are not enabled.
```

>
> [!NOTE]

> > It's highly recommended that you set the `TestingPlatformDotnetTestSupport` property in `Directory.Build.props`. That way, you don't have to add it to every test project file, and you don't risk introducing a new project that doesn't set this property and end up with a solution where some projects are VSTest while others are Microsoft.Testing.Platform, which may not work correctly and is unsupported scenario.

Reviewer 2: ```suggestion
> It is highly recommended to set the `TestingPlatformDotnetTestSupport` property in `Directory.Build.props`. This approach ensures that you don't need to add it to every test project file individually. Additionally, it prevents the risk of introducing a new project that doesn't set this property, which could result in a solution where some projects use VSTest while others use Microsoft.Testing.Platform. This mixed configuration may not work correctly and is an unsupported scenario.
```

Reviewer: may => might

> The following command-line options of `dotnet test` command in VSTest mode are supported by Microsoft.Testing.Platform. These options are build-specific and not passed down to VSTest, which is why they work well with MTP.

Reviewer 2: ```suggestion
The following command-line options for the `dotnet test` command in VSTest mode are supported by Microsoft.Testing.Platform. These options are specific to the build process and are not passed down to VSTest, which is why they work effectively with Microsoft.Testing.Platform.
```

> The list below described all `dotnet test` command line options that are supported by `Microsoft.Testing.Platform`:

Author: ```suggestion
The list below described all dotnet test command line options that are supported by Microsoft.Testing.Platform when running in VSTest mode of `dotnet test`:
```

Reviewer 2: ```suggestion
The list below outlines all the `dotnet test` command-line options that are supported by Microsoft.Testing.Platform when running in VSTest mode.
```

Reviewer: ```suggestion
The following list outlines all the `dotnet test` command-line options that are supported by Microsoft.Testing.Platform when running in VSTest mode.
```

- `-a|--arch <ARCHITECTURE>`
- `--artifacts-path <ARTIFACTS_DIR>`
- `-c|--configuration <CONFIGURATION>`
- `-f|--framework <FRAMEWORK>`

> - `-e|--environment <NAME="VALUE">`

Reviewer: Does the value have to be in quotes?

Author: It depends. If it contains spaces or semicolons, then double quotes are required. This is basically originating from the `dotnet test` CLI reference which was added in https://github.com/dotnet/docs/pull/30295 and remained as-is since then.

>
> [!IMPORTANT]
> If your test project supports VSTest but does not support MTP, an error will be generated.
>
> [!TIP]

> > You can customize the command-line arguments of your test application via `TestingPlatformCommandLineArguments` MSBuild property:

Reviewer: This seems to be repeated from line 57.

Author: @gewarren Yes. It's intentional. One is under the VSTest mode of `dotnet test` and the other is under the MTP mode of `dotnet test`. The property is supported in both cases.
